rust/src/rest.rs

```diff
@@ -0,0 +1,114 @@
+use crate::service::Handle;
+use crate::ParticipantPublicKey;
+use crate::crypto::ByteObject;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use bytes::Bytes;
+use bytes::Buf;
+use warp::Filter;
+use warp::reject::Rejection;
+use std::convert::Infallible;
+use warp::http::{StatusCode, Response};
+
+async fn handle_whatever(name: String) -> Result<impl warp::Reply, Rejection> {
+    match name.len() {
+        2 => Ok(warp::reply()),
+        _default => Err(warp::reject::not_found()),
+    }
+
+}
+
+pub async fn serve(addr: impl Into<SocketAddr> + 'static, handle: Handle) {
+    let route = warp::path!("hello" / String)
+//      .map(|name| format!("Hello, {}!", name));
+        .and_then(handle_whatever);
+
+    let message = warp::path!("message")
+        .and(warp::post())
+        .and(warp::body::bytes())
+        .and(with_hdl(handle.clone()))
+        .and_then(handle_message);
+
+    let sum_dict = warp::path!("sums")
+        .and(warp::get())
+        .and(with_hdl(handle.clone()))
+        .and_then(handle_sums);
+
+    // can't take pk as param in the path, as it doesn't impl FromStr...
+    let seed_dict = warp::path!("seeds")
+        .and(warp::get())
+        // ... so let's assume that it's in the request body
+        // it's not good HTTP 1.1 practice but what can you do
+        .and(warp::body::bytes())
+        .and(with_hdl(handle.clone()))
+        .and_then(handle_seeds);
+
+    let round_params = warp::path!("params")
+        .and(warp::get())
+        .and(with_hdl(handle))
+        .and_then(handle_params);
+
+    let _routes = message
+        .or(sum_dict)
+        .or(seed_dict)
+        .or(round_params);
+
+    warp::serve(route).run(addr).await
+}
+
+async fn handle_message(body: Bytes, handle: Handle) -> Result<impl warp::Reply, Infallible> {
+    handle.send_message(body.to_vec()).await;
+    Ok(warp::reply())
+}
+
+async fn handle_sums(handle: Handle) -> Result<impl warp::Reply, Rejection> {
+    let sum_dict = handle
+        .get_sum_dict()
+        .await
+        .ok_or(warp::reject::not_found())?;
+    let sum_dict_bytes = Arc::try_unwrap(sum_dict).unwrap(); // HACK
```

Comment:
this will panic if more than one clone of the `Arc<Vec<_>>` is circulating. dereferencing the arc probably doesn't work since the inner `Vec<_>` is not `Copy`. we might check against `Arc::strong_count() == 1` to safely unwrap or we might have to clone if `with_header()` doesn't accept references.

Reply:
again that was an older version containing a hack, but thanks for the suggestion!